### ISSUE TITLE
LoRA training memory efficiency: merge-and-reload approach

### ISSUE THREAD

**Issue by schancel** [darwin-finch/finch#7]
## Overview

The current LoRA training architecture has high memory overhead because PyTorch must load the model a second time to apply gradients. This issue tracks improving the training pipeline's memory efficiency.

## Current Architecture

```
Inference:  ONNX Runtime loads base model (e.g., 3GB for Qwen-2.5-3B)
Training:   Python script loads same model via PyTorch (another 3GB)
Peak:       6GB+ memory usage during training
```

This is acceptable on 16GB+ machines but problematic on 8GB machines.

## Proposed Improvement: Merge-and-Reload Pipeline

```
1. Collect feedback examples → ~/.finch/feedback.jsonl
2. When batch is ready, trigger Python merge script:
   a. Load base model via PyTorch
   b. Apply LoRA fine-tuning (or load existing adapter)
   c. Merge LoRA weights into base model
   d. Export merged model as new ONNX file → ~/.finch/models/merged_v2.onnx
3. Daemon hot-reloads merged ONNX (swap pointer, drain in-flight requests)
4. Delete old ONNX file
```

Peak memory during step 2 is still 2x, but training runs in a separate process 
(doesn't block inference) and the merged model replaces the original (no ongoing overhead).

## Alternative: Pure Rust LoRA (Long-term)

Implement LoRA as a thin Rust layer on top of ONNX Runtime:
- Load base model weights + LoRA adapter weights separately
- Apply LoRA during each forward pass: `W_effective = W_base + alpha * (A @ B)`
- No Python dependency, no model duplication
- Estimated effort: 40-80 hours

## Files

- `scripts/train_lora.py` — add merge-and-export mode
- `src/training/batch_trainer.rs` — trigger merge after training
- `src/models/loaders/onnx.rs` — hot-reload support
- `src/models/lora.rs` — future Rust LoRA layer

## Acceptance Criteria (Merge-and-Reload)

- [ ] `scripts/merge_lora.py` merges adapter + base model → new ONNX
- [ ] Daemon detects new ONNX file and reloads without dropping connections
- [ ] Memory during merge stays under 2x base model size
- [ ] Merged model produces better outputs for trained patterns

**Comment by schancel** [comment]
2026-08-25 audit: this substantially duplicates #1. Retain only a measured peak-memory optimization objective after #1's functional merge/reload path exists, or close as duplicate. Do not build a second adapter-loading design.

**Comment by schancel** [comment]
2026-08-25 priority update: blocked behind #1 and #74. Do not implement the proposed Python merge-and-reload bulk-training pipeline. Current ONNX Runtime supports on-device training APIs and GenAI adapter inference, but its documented LoRA training/conversion path still depends on Python/Olive/PyTorch and does not establish a practical native Finch/macOS/Rust pipeline. Reopen this only after #1 has a verified non-Python functional path; then retain #7 solely for measured peak-memory optimization rather than a second adapter design. Primary references: https://onnxruntime.ai/docs/get-started/training-on-device.html and https://onnxruntime.ai/docs/genai/tutorials/finetune.html

**Comment by schancel** [comment]
Queue: value 1, cost 2, certainty 2, unblocking 0; score 1.0. Substantially duplicates #1's merge/reload outcome, and automatic training is disabled pending a supported path. Not ready: duplicate of #1; blocked/deferred
